dubbo-common/src/main/java/org/apache/dubbo/common/utils/JVMUtil.java

```diff
@@ -33,75 +33,84 @@ public class JVMUtil {
     public static void jstack(OutputStream stream) throws Exception {
         ThreadMXBean threadMxBean = ManagementFactory.getThreadMXBean();
         for (ThreadInfo threadInfo : threadMxBean.dumpAllThreads(true, true)) {
-            stream.write(getThreadDumpString(threadInfo).getBytes());
+            getThreadDumpString(stream, threadInfo);
         }
     }
 
-    private static String getThreadDumpString(ThreadInfo threadInfo) {
-        StringBuilder sb = new StringBuilder("\"" + threadInfo.getThreadName() + "\"" + " Id="
-                + threadInfo.getThreadId() + " " + threadInfo.getThreadState());
+    private static void getThreadDumpString(final OutputStream stream, ThreadInfo threadInfo) throws Exception {
+        // print basic info
+        stream.write(String.format(
+                        "\"%s\" Id=%d %s",
+                        threadInfo.getThreadName(), threadInfo.getThreadId(), threadInfo.getThreadState())
+                .getBytes());
         if (threadInfo.getLockName() != null) {
-            sb.append(" on " + threadInfo.getLockName());
+            stream.write(String.format(" on %s", threadInfo.getLockName()).getBytes());
         }
         if (threadInfo.getLockOwnerName() != null) {
-            sb.append(" owned by \"" + threadInfo.getLockOwnerName() + "\" Id=" + threadInfo.getLockOwnerId());
+            stream.write(
+                    String.format(" owned by \"%s\" Id=%d", threadInfo.getLockOwnerName(), threadInfo.getLockOwnerId())
+                            .getBytes());
         }
         if (threadInfo.isSuspended()) {
-            sb.append(" (suspended)");
+            stream.write(" (suspended)".getBytes());
         }
         if (threadInfo.isInNative()) {
-            sb.append(" (in native)");
+            stream.write(" (in native)".getBytes());
         }
-        sb.append('\n');
-        int i = 0;
+        stream.write("\n".getBytes());
+
+        // calculate stack print depth
+        StackTraceElement[] stackTrace = threadInfo.getStackTrace();
         // default is 32, means only print up to 32 lines
-        int jstackMaxLine = 32;
+        int printStackDepth = Math.min(32, stackTrace.length);
         String jstackMaxLineStr = System.getProperty(CommonConstants.DUBBO_JSTACK_MAXLINE);
         if (StringUtils.isNotEmpty(jstackMaxLineStr)) {
             try {
-                jstackMaxLine = Integer.parseInt(jstackMaxLineStr);
+                printStackDepth = Integer.parseInt(jstackMaxLineStr);
```

Comment:
Seems should be `Math.min(stackTrace.length, Integer.parseInt(jstackMaxLineStr))`